bug/pager - DISMISS_EDITOR, REMOVE_ROW, and ADD_NEW_ROW now set total correctly

Vanderslice opened the pull request:
Fix for bug #132.

Feedback Requested: I moved the logic for updating total from addNewRows and dismissEditor to saveRow so the grid totals update on save. I can see arguments for either.

Note: tests are currently failing due to moving the logic, will update after getting feedback

bencripps commented:
I think these changes look good 👍 

Let's get the tests fixed; I think this is a good/sensible change.

Vanderslice commented:
I went ahead and moved the logic for updating the totals back to addNewRow and dismissEditor for the time being due to a conflict with the latest changes to allow for inserting a new row editor at any point in the grid

bencripps commented:
Changes look good. Sort of unrelated, how come we don't update the total in `removeRow` like we do with `addRow`? Here's the snippet from the actionHandler for `removeRow`:

```js
const updated = record.merge({
        data: remainingRows,
        proxy: remainingRows,
        currentRecords: remainingRows,
        lastUpdate: generateLastUpdate()
    });
```

Should we add `total` to this update?

bencripps commented on src/reducers/actionHelpers/datasource.js:
```diff
@@ -165,8 +162,8 @@ export const addNewRow = (state, { rowId, stateKey }) => {
         isEditing: true,
-        lastUpdate: generateLastUpdate(),
-        total: newData.size
+        total: existingState.get('total') + 1,
```
Probably need to be a little bit more defensive here sadly. I would add a check to make sure existingState has been initialized, otherwise this will throw

Vanderslice commented:
Thanks for the feedback! I updated the removeRow actionHandler to decrement the total and fixed the tests to handle this change. 

I had previously missed this because we do a setData after removing rows but I could see where this may not be wanted/needed plus it makes the actionHandlers behave consistently. 

Deleting rows on the demo grid now gives you instant feedback that the number of currently displayed rows and total have changed without requiring setData/refresh

bencripps commented on src/reducers/actionHelpers/datasource.js:
```diff
@@ -96,16 +95,24 @@ export const dismissEditor = (state, { stateKey }) => {
     const record = state.get(stateKey);
 
+    const updatedTotal = existingState
+    && existingState.get('total')
```
Run `npm run lint` and see if you get any errors

bencripps commented on src/reducers/actionHelpers/datasource.js:
```diff
@@ -96,16 +95,24 @@ export const dismissEditor = (state, { stateKey }) => {
         data: remainingRows,
         proxy: remainingRows,
         currentRecords: remainingRows,
+        total: updatedTotal - 1,
```
This is a nit, but if total resolved to 0, then we're setting the total to `-1` which probably isn't best.
